instance/models/mixins/storage.py

```diff
@@ -211,14 +212,30 @@ def get_s3_connection(self):
             self.s3_secret_access_key
         )
 
-    def provision_s3(self):
+    def _create_bucket(self, attempts=4, attempt=1, retry_delay=1):
         """
-        Create S3 Bucket if it doesn't exist
+        Create bucket, retry up to defined attempts if it fails
         """
-        if self.s3_access_key and self.s3_secret_access_key:
+        try:
             s3 = self.get_s3_connection()
             bucket = s3.create_bucket(self.s3_bucket_name)
             bucket.set_cors(get_s3_cors_config())
+        except boto.exception.S3ResponseError:
+            if attempts == attempt:
+                raise
+            self.logger.exception(
+                'Retrying bucket creation. IAM keys are not propagated yet, attempt %s of %s',
+                attempt, attempts
+            )
+            time.sleep(retry_delay)
+            self._create_bucket(attempt + 1)
```

Comment:
This will pass `2` into `attempts` on the first retry, rather than update `attempt`. Can be fixed by specifying `attempt` as the key for this value.

However you can probably do without a separate `attempt` key. Just pass in `attempts - 1`, which will take the 4 to a 3, and check if `attempts == 0` to decide to re-raise the error. Maybe made easier to understand by using the name `attempts_remaining`.

Reply:
Good catch, just fixed it. Thanks!